src/sidebar/util/view-filter.js

```diff
@@ -0,0 +1,195 @@
+ */
+
+/**
+ * @typedef {import('../../types/api').Annotation} Annotation
+ */
+
+import { quote } from './annotation-metadata';
+import * as unicodeUtils from './unicode';
+
+/**
+ * @typedef Filter
+ * @prop {(ann: Annotation) => boolean} matches
+ */
+
+/**
+ * @typedef Checker
+ * @prop {(ann: Annotation) => boolean} autofalse
+ * @prop {(ann: Annotation) => any|any[]} value
+ * @prop {(term: any, value: any) => boolean} match
+ */
+
+function displayName(ann) {
+  if (!ann.user_info) {
+    return '';
+  }
+  return ann.user_info.display_name || '';
+}
+
+/**
+ * Normalize a field value or query term for comparison.
+ */
+function normalize(val) {
+  if (typeof val !== 'string') {
+    return val;
+  }
+  return unicodeUtils.fold(unicodeUtils.normalize(val)).toLowerCase();
+}
+
+/**
+ * Filter that matches annotations against a single field & term.
+ *
+ * eg. "quote:foo" or "text:bar"
+ *
+ * @implements {Filter}
+ */
+class TermFilter {
+  /**
+   * @param {string} field - Name of field to match
+   * @param {string} term - Query term
+   * @param {Checker} checker - Functions for extracting term values from
+   *   an annotation and checking whether they match a query term.
+   */
+  constructor(field, term, checker) {
+    this.field = field;
+    this.term = term;
+    this.checker = checker;
+  }
+
+  matches(ann) {
+    const checker = this.checker;
+    if (checker.autofalse && checker.autofalse(ann)) {
+      return false;
+    }
+
+    let value = checker.value(ann);
+    if (Array.isArray(value)) {
+      value = value.map(normalize);
+    } else {
+      value = normalize(value);
+    }
+    return checker.match(this.term, value);
+  }
+}
+
+/**
+ * Filter that combines other filters using AND or OR combinators.
+ *
+ * @implements {Filter}
+ */
+class BinaryOpFilter {
+  /**
+   * @param {'and'|'or'} op - Binary operator
+   * @param {Filter[]} filters - Array of filters to test against
+   */
+  constructor(op, filters) {
+    this.operator = op;
+    this.filters = filters;
+  }
+
+  matches(ann) {
+    if (this.operator === 'and') {
+      return this.filters.every(filter => filter.matches(ann));
+    } else {
+      return this.filters.some(filter => filter.matches(ann));
+    }
+  }
+}
+
+/**
+ * Filters a set of annotations.
+ *
+ * @param {Annotation[]} annotations
+ * @param {Object} filters - Faceted filter generated by
+ * `generateFacetedFilter`.
+ * @return {string[]} IDs of matching annotations.
+ */
+export default function viewFilter(annotations, filters) {
+  /**
+   * Functions for extracting field values from annotations and testing whether
+   * they match a query term.
+   *
+   * [facet_name]:
+   *   autofalse: a function for a preliminary false match result
+   *   value: a function to extract to facet value for the annotation.
+   *   match: a function to check if the extracted value matches the facet value
+   *
+   * @type {Object.<string,Checker>}
+   */
+  const fieldMatchers = {
```

Comment:
I think `fieldMatchers` could be hoisted to the top level of the module to avoid recreating it on each call, since it doesn't depend on any of the function parameters.

Reply:
Indeed.